library/hashmap.lisp

```diff
@@ -484,7 +674,95 @@ removed.  If HM does not contain an entry with KEY, HM is returned as is."
     "Returns an interator to iterate over all the values in a hashmap hm."
     (iter:new (->generator hm (fn (_ v) v))))
 
-  ;; Debug tools
+  ;; API
+  (declare entries (Hash :k => HashMap :k :v -> (iter:Iterator (Tuple :k :v))))
+  (define (entries hm)
+    "Returns an interator to iterate over all entries in hashmap hm."
+    (iter:new (->generator hm Tuple)))
+  )
+
+;;
+;; Instances
+;;
+
+(coalton-toplevel
+  (define-instance (iter:IntoIterator (HashMap :k :v) (Tuple :k :v))
+    (define (iter:into-iter hm)
+      (iter:new (->generator hm Tuple))))
+
+  (define-instance ((Eq :k) (Eq :v) (Hash :k) => Eq (HashMap :k :v))
+    (define (== a b)
+      (and (== (count a) (count b))
+           (iter:every! (fn ((Tuple k v))
+                          (== (lookup b k) (Some v)))
+                        (entries a)))))
+
+  (define-instance ((Hash :k) (Hash :v) => Hash (HashMap :k :v))
+    (define (hash hm)
+      (iter:elementwise-hash! (iter:into-iter hm))))
+
+  (define-instance (Hash :k => iter:FromIterator (HashMap :k :v) (Tuple :k :v))
+    (define iter:collect! collect!))
+
+  (define-instance (Functor (HashMap :key))
+    (define (map func mp)
+      (HashMap (%map func (root mp)))))
+  )
+
+;;
+;; Set operations
+;;
+
+(coalton-toplevel
+
+  (declare union (Hash :k => HashMap :k :v -> HashMap :k :v -> HashMap :k :v))
+  (define (union a b)
+    "Construct a HashMap containing all the mappings from A and B.
+
+If A and B contain mappings X -> A' and X -> B', the former mapping is kept.
+
+Because of the possibility that A and B will map the same X to different A' and B', this is not an associative
+operation, and therefore OrdMap cannot implement Monoid."
```

Comment:
Should be HashMap in the comment

Reply:
Wait, is this true? It looks like HashMap implements SemiGroup below. Is that instance not lawful? Or is this comment just wrong?

Reply:
Actually I lifted that doc from original OrdMap, and I see is wrong now, since the precedence in case of key conflict is now clarified and `uinon` is associative (but not commutative).   It can also form a Monoid, with an empty map as an identity.

Whether adoping `union` to define `<>` is debatable, though.  Technically `inetrsection` is also associative.